### ISSUE TITLE
[Spell] Lifebloom

### ISSUE THREAD

**Issue by Hajzlopuc** [WoWManiaUK/Redemption#2232]
Hello,

there is an issue how lifebloom's last tick is being calculated.
http://www.wow-mania.com/armory/?spell=33778 this should be the spell ID of final bloom.

Also spell doesn't benefit properly from talents:
[Gift of Nature](http://www.wow-mania.com/armory/?spell=24946)
[Empowered Rejuvenation](http://www.wow-mania.com/armory/?spell=33890)

Right now, the bloom (last tick) is being calculated like this:
Only the first stack benefits from healing power bonus.
for 1 stack, bloom - 776 + (healing bonus * 0.284) ---> Though that 0.284 is also not correct according to wowwikki and other sources, but that's not the core of the issue, so I will use that in all further formulas.
The issue is this:
for 2 stack, bloom - 776 + 776 + (healing bonus * 0.284)
for 3 stack, bloom - 776 + 776 + 776 + (healing bonus * 0.284)

This should be:
for 2 stack, bloom = (776 + (healing bonus * 0.284)) + (776 + (healing bonus * 0.284))
for 3 stack, bloom = (776 + (healing bonus * 0.284)) + (776 + (healing bonus * 0.284)) + (776 + (healing bonus * 0.284))

I have tested it in more scenarios:

**1st case** - 0 spell power, 0 talents
https://puu.sh/zkbwz/cef2be32c7.jpg
1 stack - 776
2 stacks - 776+776=1552
3 stacks - 776+776+776=2328
This is correct.

**2nd case** - 3157 spell power, 0 talents
https://puu.sh/zkbA0/c44c14dfd4.jpg
1 stack - 1672 
776 + (3157*0.284) = 1672
this one is correct for 1 stack

but

2 stacks - 2448
776 + 776 + (3157 * 0.284) = 2448
this should be
(776 + (3157 * 0.284)) + (776 + (3157 * 0.284)) = 3344

3 stacks - 3224
776 + 776 + 776 + (3157 * 0.284) = 3224
and should be
(776 + (3157 * 0.284)) + (776 + (3157 * 0.284)) + (776 + (3157 * 0.284)) = 5016

**3rd case** - 3157 spell power and 5/5 Gift of Nature talent
https://puu.sh/zkbCz/cfca83d119.jpg
http://wowwiki.wikia.com/wiki/Gift_of_Nature
According to wowwiki - For calculation purposes, this talent increases the healing of each spell **after** spell power is applied.

Right now the talent only applies to the base value of 776

1 stack - 1749
(776 * 1.1) + (3157 * 0.284) = 1749
this should be
(776 + (3157 * 0.284)) * 1.1 = 1839
2 stacks - 2603
(776 + 776) * 1.1 + (3157 * 0.284) = 2603
should be
((776 + (3157 * 0.284)) + (776 + (3157 * 0.284))) * 1.1 = 3678
3 stacks - 3456
(776 + 776 + 776) * 1.1 + (3157 * 0.284) = 3456
this should be
((776 + (3157 * 0.284)) + (776 + (3157 * 0.284)) + (776 + (3157 * 0.284))) * 1.1 = 5517

Furthermore
**4th case** - 3157 spell power, 5/5 talents in Gift of Nature and 5/5 talents in Empowered rejuvenation
https://puu.sh/zkbGc/49c0fc83a1.jpg
http://wowwiki.wikia.com/wiki/Empowered_Rejuvenation
According to wiki since patch 2.4.0 - Empowered Rejuvenation applies to both the HoT **and the direct heal portions of Regrowth and Lifebloom.** 

As you see on the screenshot, it returns the same exact value as case number 3 which is without this talent.

To sum this up:
1. Lifebloom's bloom effect with 2 or 3 stacks does not benefit from spell power. Spell power is only applied to first stack. (Also the spell power coefficient seems to be bad.)
2. Lifebloom's bloom effect does not benefit fully from Gift of Nature talent. Only the base value 776 benefits from the talent.
3. Lifebloom's bloom effect does not benefit at all from the talent Empowered Rejuvenation

I am pretty sure that few days ago Lifebloom worked without any issue. However I'm not sure since when it's broken, I noticed only today.

**Comment by fvicaria** [comment]
I will have a look into it tomorrow.

**Comment by Hajzlopuc** [comment]
For spell power coefficient I have found these on mangos old forums.
https://puu.sh/zkhwT/116a4464b6.jpg

Then patch told us: Patch 3.2.0 (04-Aug-2009): The final heal that occurs when this spell blooms has been reduced by 20% on the base and on the spell power coefficient.

https://puu.sh/zkhx1/ef5ee6bb3e.jpg
Seems to be correct. It's been nerfed by 20% in patch, that means: 970 * 0.8 =776 which is the base value now and 64.6 * 0.8 = 51.68 which should be the coefficient now.
Also guys from trinity core used this coefficient recently.
https://github.com/TrinityCore/TrinityCore/issues/21297#issuecomment-359808981

About Empowered Rejuvenation. 
Patch 2.4.0 (25-Mar-2008): "Empowered Rejuvenation": This talent now properly affects the final heal from Lifebloom. 
And it seems that nothing have ever changed since then.
Also this might help.
https://github.com/TrinityCore/TrinityCore/issues/244

**Comment by Hajzlopuc** [comment]
Tested it on dev realm, still not fixed.
Let's look at the core of the issue.

879 was my spell power and 0.283 was the spell coefficient I calculated which is also incorrect but let's start with the core of the issue first
https://puu.sh/zldk8/af5e8913d7.jpg

1 stack healed me for 1025 like this 
776 * 1 + (879 * 0.283) = 1025
should be:
(776 + (879 * 0.283)) * 1 = 1025

2 stacks healed me for 1801 like this:
776 * 2 + (879 * 0.283) = 1801
should be:
(776 + (879 * 0.283)) * 2 = 2050

3 stacks healed me for 2577 like this
776 * 3 + (879 * 0.283) = 2577
should be:
(776 + (879 * 0.283)) * 3 = 3075

the problem is that the stack multiplier is only applied on base balue of 776 at the moment

**Comment by fvicaria** [comment]
Can someone confirm this?

**Comment by Dranah** [comment]
I just tested, and can confirm the problem with the spell coeff not stacking for the burst part as intended, as well as the coefficients being entirely wrong.
I tested without talents to ignore any other effects with a bonus healing of 580.

1. Spell power coefficient of HOT portion (ticks):
Base healing is 53 / tick.

How it should work: each tick should receive 6,53% of your bonus healing (53 + 38 = 91 is expected value)
How it works right now: heals for 108 / tick instead. (+55, which means 9,48% coefficient)
This coefficient stacks properly, the ticks increase to 216/324.

2. Spell power coefficient and stacking of Burst portion (bloom):
Base healing is 776 per stack.

How it should work: It should receive 38,57% of your bonus healing ( 776+ 224 =1000 is expected value for me)
Stacking Lifebloom should multiply this healing by 2 or 3 accordingly.

How it works right now: heals for 940 on the bloom, which is 28,27% coefficient that Hajz also experienced.
This coefficient does not stack as the Lifebloom is stacked, each further stack gives only + 776 healing which is the base value.

3. ICC healing buff: "Hellscream's Warsong"

How it should work: increases healing effects by 30% - this is the part which affects Lifebloom.
This means, testing with 0 bonus healing, the base values should change as
53 x 1,3 = 68 healing
776 x 1,3 = 1008 healing

How it works atm: double dips the bloom portion, multiplicatively, increasing it's healing by 30% and then 30% again.
Ticks work as intended: 68 healing per tick.
Burst: 776 x 1,3 = 1008 , 1008x 1,3 = 1310
It is currently healing for 1310. (+68,8%)

When testing with spell power, the results are similar, but anomalic, the numbers are a bit off even the bugged values, probably due to the already wrong spell formula. I can't really figure how it calculates at this point as multiple parts of the equation are wrong.
Example: 
The intended values would be 1000 outside icc and 1300 inside icc, with my bonus healing and 1 stack. 2000/2600 with two stacks, and 3000/3900 with 3 stacks.
How it works at the moment: 1 stack blooms for 940 outside ICC as stated above. Inside icc, it blooms for 1523, which is +62%, not +68,8% - but still way too much, even with the broken SP coeff.
2 stacks heal for 1716 outside icc , 2835 inside icc (+65,2% there)
3 stacks heal 2492 outside icc, 4147 inside icc (+66,4% there)
Result is, even with the spell coeff being too low, and not stacking on the bloom, the bloom ends up healing more than it should when inside ICC, because the buff boosts it too much. Once the spell coeff and stacking are fixed, this would result in lifebloom being damn OP in icc.

**Comment by fvicaria** [comment]
Please test again.

**Comment by Dranah** [comment]
Everything is the same as before. Coefficients, stacking behavior and ICC scaling.

**Comment by Dranah** [comment]
UPDATE:

1. Spell power coefficients are still wrong. (tick portion 9,48 instead of 6,53 / bloom portion 28,27 instead of 38,57) (need fix)
2. Stacking of burst healing value is still incorrect. Only stacks the base healing. (need fix)
3. Overtuned scaling with ICC buff is fixed. Both ticks and blooms now receive 30%. (fixed)

4. Gift of Nature http://www.wow-mania.com/armory/?spell=24946
Additionally, I just tested out the "Gift of Nature" talent as well (completely missed it previously):
Well I have no idea if any work was done on this, but right now it scales with the burst part properly.
1 stack: 940 x 1,1 = 1035
2 stack 1716 x 1,1 = 1888 
3 stack 2493 x 1,1 = 2742
The values it multiplies are wrong by default, but the 10% modifier itself works on them as intended.

**However**, it's effect on the HoT portion (more precisely, on HOTs with spell bonus) is wrong:
- Test with no spell power: tick is properly increased 53 x 1,1 = 58 for 1 stack, 106 x 1,1 = 116 for 2 stack etc. (1 stack of lifebloom)
- Test with 580 spell power: tick is increased from 108 to 124. Expected value is 108 x 1,1 = 119 (1 stack of lifebloom)

This is created most likely by the following problem: the talent increases the base healing from 53 to 58, then adds the bonus spell healing, which makes it 113, then multiplies the entire healing by 10% again and 113 x 1,1 is 124.
Same goes on 2 stacks: 216 x 1,1 = 238 is expected value. Instead it ticks for 249, which is created by base healing 2x53 x 1,1 = 117 + bonus healing 2x 55 makes it 227 and 227 x 1,1 =249.

**Comment by Dranah** [comment]
- spell power coefficient of bloom has shifted to 35,9% (was 28,3, should be 38,5)
Getting there.

**Comment by fvicaria** [comment]
It will be left "as-is"
